genesis/engine/couplers/sap_coupler.py

```diff
@@ -1740,3 +2393,402 @@ def compute_delassus(self, i_p):
         # With floor, J is Identity
         W = self.fem_solver.pcg_state_v[i_b, i_g].prec
         return W
+
+
+@ti.data_oriented
+class RigidFloorVertContact(RigidContact):
+    def __init__(
+        self,
+        simulator: "Simulator",
+    ) -> None:
+        super().__init__(simulator)
+        self.name = "RigidFloorVertContact"
+        self.rigid_solver = self.sim.rigid_solver
+        self.floor_height = self.sim.fem_solver.floor_height
+        self.contact_pair_type = ti.types.struct(
+            batch_idx=gs.ti_int,  # batch index
+            geom_idx=gs.ti_int,  # index of the vertex
+            link_idx=gs.ti_int,  # index of the link
+            contact_pos=gs.ti_vec3,  # contact position
+            sap_info=self.sap_contact_info_type,  # contact info
+        )
+        self.max_contact_pairs = self.rigid_solver.n_free_verts * self.sim._B
+        self.contact_pairs = self.contact_pair_type.field(shape=(self.max_contact_pairs,))
+        self.Jt = ti.field(gs.ti_vec3, shape=(self.max_contact_pairs, self.rigid_solver.n_dofs))
+        self.M_inv_Jt = ti.field(gs.ti_vec3, shape=(self.max_contact_pairs, self.rigid_solver.n_dofs))
+        self.W = ti.field(gs.ti_mat3, shape=(self.max_contact_pairs,))
+
+    @ti.func
+    def detection(self, f: ti.i32):
+        overflow = False
+        pairs = ti.static(self.contact_pairs)
+        sap_info = ti.static(pairs.sap_info)
+        C = ti.static(1.0e6)
+        # Compute contact pairs
+        self.n_contact_pairs[None] = 0
+        for i_b, i_v in ti.ndrange(self.rigid_solver._B, self.rigid_solver.n_verts):
+            if not self.rigid_solver.verts_info.is_free[i_v]:
+                continue
+            i_fv = self.rigid_solver.verts_info.verts_state_idx[i_v]
+            pos_v = self.rigid_solver.free_verts_state.pos[i_fv, i_b]
+            distance = pos_v.z - self.floor_height
+            if distance > 0:
+                continue
+            i_g = self.rigid_solver.verts_info.geom_idx[i_v]
+            i_l = self.rigid_solver.geoms_info.link_idx[i_g]
+            i_p = ti.atomic_add(self.n_contact_pairs[None], 1)
+            if i_p < self.max_contact_pairs:
+                pairs[i_p].batch_idx = i_b
+                pairs[i_p].geom_idx = i_fv
+                pairs[i_p].link_idx = i_l
+                pairs[i_p].contact_pos = pos_v
+                sap_info[i_p].k = C
+                sap_info[i_p].phi0 = distance
+                sap_info[i_p].mu = self.rigid_solver.geoms_info.coup_friction[i_g]
+            else:
+                overflow = True
+        return overflow
+
+    @ti.func
+    def compute_W(self):
```

Comment:
Use a better name and add short docstring to explain what this quantity is if not directly obvious.